w3lib/url.py

```diff
@@ -422,7 +422,9 @@ def parse_url(url, encoding=None):
     parsed url)
     """
     if isinstance(url, ParseResult):
-        return url
+        # we need to make sure we have unicode parts
+        # otherwise, things like IDNA encoding may fail on bytes
+        return ParseResult(*(to_unicode(part, encoding) for part in url))
```

Comment:
It seems the root issue is that IDNA encoding is defined only for unicode, not for bytes. So we should convert netloc to unicode before encoding it in _safe_ParseResult.